src/com/owncloud/android/files/services/FileUploadService.java

```diff
@@ -0,0 +1,1359 @@
+        return account.name + remotePath;
+    }
+
+    private String buildRemoteName(UploadDbObject uploadDbObject) {
+        return uploadDbObject.getAccountName() + uploadDbObject.getRemotePath();
+    }
+
+    /**
+     * Checks if an ownCloud server version should support chunked uploads.
+     * 
+     * @param version OwnCloud version instance corresponding to an ownCloud
+     *            server.
+     * @return 'True' if the ownCloud server with version supports chunked
+     *         uploads.
+     */
+    private static boolean chunkedUploadIsSupported(OwnCloudVersion version) {
+        return (version != null && version.compareTo(OwnCloudVersion.owncloud_v4_5) >= 0);
+    }
+
+    /**
+     * Service initialization
+     */
+    @Override
+    public void onCreate() {
+        super.onCreate();
+        Log_OC.d(TAG, "mPendingUploads size:" + mPendingUploads.size() + " - onCreate");
+        mNotificationManager = (NotificationManager) getSystemService(NOTIFICATION_SERVICE);
+        mBinder = new FileUploaderBinder();
+        mDb = UploadDbHandler.getInstance(this.getBaseContext());
+        mDb.recreateDb(); //for testing only
+        
+        //when this service starts there is no upload in progress. if db says so, app probably crashed before.
+        mDb.setAllCurrentToUploadLater();
+        
+        HandlerThread thread = new HandlerThread("FileUploadService-Requester");
+        thread.start();
+        
+        mServiceLooper = thread.getLooper();
+        mServiceHandler = new ServiceHandler(mServiceLooper);
+
+        uploadExecutor = Executors.newFixedThreadPool(1);
+
+        Log_OC.d(TAG, "FileUploadService.retry() called by onCreate()");
+        FileUploadService.retry(getApplicationContext());
+    }
+    
+    @Override
+    public void onDestroy() {
+        Log_OC.d(TAG, "mPendingUploads size:" + mPendingUploads.size() + " - onDestroy");
+        mServiceLooper.quit();
+        uploadExecutor.shutdown();
+    }
+
+    private final class ServiceHandler extends Handler {
+        public ServiceHandler(Looper looper) {
+            super(looper);
+        }
+
+        @Override
+        public void handleMessage(Message msg) {
+            onHandleIntent((Intent) msg.obj, (int)msg.arg1);
+        }
+    }
+
+    @Override
+    public void onStart(Intent intent, int intentStartId) {
+        Message msg = mServiceHandler.obtainMessage();
+        msg.arg1 = intentStartId;
+        msg.obj = intent;
+        mServiceHandler.sendMessage(msg);
+    }
+
+    @Override
+    public int onStartCommand(Intent intent, int flags, int startId) {
+        onStart(intent, startId);
+        return START_STICKY; // if service is killed by OS before calling
+                             // stopSelf(), tell OS to restart service with
+                             // null-intent. 
+    }
+
+    /**
+     * The IntentService calls this method from the default worker thread with
```

Comment:
Oh crap, yes. For some reason I do not recall, I re-implemented IntentService here. It might be because it wanted to get rid of the ``stopSelf()`` in the `ServiceHandler` --> http://grepcode.com/file/repo1.maven.org/maven2/org.robolectric/android-all/4.4_r1-robolectric-1/android/app/IntentService.java#66 but this is surely overkill.

BTW, this also explains the `volatile` keywords and the `Executor` above: They are taken from `IntentService`.